src/main/java/org/springframework/cloud/bindings/boot/MySqlBindingsPropertiesProcessor.java

```diff
@@ -68,12 +78,34 @@ public void process(Environment environment, Bindings bindings, Map<String, Obje
             //r2dbc properties
             map.from("password").to("spring.r2dbc.password");
             map.from("host", "port", "database").to("spring.r2dbc.url",
-                    (host, port, database) -> String.format("r2dbc:mysql://%s:%s/%s", host, port, database));
+                    (host, port, database) -> String.format("r2dbc:%s://%s:%s/%s", evalR2DBCProtocol(), host, port, database));
             map.from("username").to("spring.r2dbc.username");
 
             // r2dbcURL takes precedence
             map.from("r2dbc-url").to("spring.r2dbc.url");
         });
     }
+    
+    private String evalR2DBCProtocol()
+    {
+    	// Default to "mysql"
+    	String connectionProtocol = MYSQL_PROTOCOL;
+    	
+    	/* Starting with Spring Boot 2.7.0, the previous MySQL r2dbc driver is no longer supported and 
+    	 * documentation suggests using the MariaDB R2DBC driver as an alternative.  Some versions
+    	 * of the MariaDB R2DBC driver do not support "mysql" as part of the connection
+    	 * protocol; "mariadb" should be used instead when the MariaDB R2DBC driver class is on
+    	 * the classpath.
+    	 */
+    	
+    	try {
+    		Class.forName("org.mariadb.r2dbc.MariadbConnection");
+    		connectionProtocol = MARIADB_PROTOCOL;
+    	}
+    	catch (ClassNotFoundException ignored) {
+        }
```

Comment:
My hesitation is that this is hard to test because it relies on something being present/not present on the classpath.

I would feel better with something like requiring the service binding to specify the DB protocol instead, and we just modify to map the include what the binding sets, with a default of MySQL.

Would you be OK with that?

Reply:
@dmikusa-pivotal 

A few thoughts:

1. The processor has already set precedent of checking for the presence of something on the classpath
```
try {
  Class.forName("org.mariadb.jdbc.Driver", false, getClass().getClassLoader());
  properties.put("spring.datasource.driver-class-name", "org.mariadb.jdbc.Driver");
} catch (ClassNotFoundException e) {
  try {
     Class.forName("com.mysql.cj.jdbc.Driver", false, getClass().getClassLoader());
     properties.put("spring.datasource.driver-class-name", "com.mysql.cj.jdbc.Driver");
  } catch (ClassNotFoundException ignored) {
```
and includes a unit test checking for the `spring.datasource.driver-class-name` by including the MariaDB library as a test dependency in the `pom.xml` file.
```        
<dependency>
        <groupId>org.mariadb.jdbc</groupId>
        <artifactId>mariadb-java-client</artifactId>
        <scope>test</scope>
</dependency>
```
```
new MySqlBindingsPropertiesProcessor().process(environment, bindings, properties);
assertThat(properties)
   .containsEntry("spring.datasource.driver-class-name", "org.mariadb.jdbc.Driver")
```
The proposed PR follows this same paradigm.

2. Setting the DB protocol `mysql` vs `mariadb` at the service binding feels like the wrong direction of control.  The service binding should NOT be aware of this level of application implementation detail.  Example:, The fact that my application went from SpringBoot 2.6.x to SpringBoot 2.7.x and requires a MariaDB vs MySQL library should not be a concern of the service binding.

3. In what field of the service binding would you put the protocol?  MySQL Operators like the VMWare Tanzu SQL only include attributes such as port, host, username/password.  Including a the DB protocol would require either the Operator to be updated or the field to added manually by another operator persona.

4. Building on to points number 2 and 3, multiple applications could very easily be consuming the same service binding, and different applications could be consuming various levels of framework libraries or could even be polygot.  Depending of the field chosen, placing something like DB protocol at a service binding configuration could have unpredictable or negative results.  For augment sake (and I'm not saying this would be the answer) let's say the service binding provided the `r2dbc-url` field which included the 'mariadb' protocol in the URL.  This would probably fix the SpringBoot 2.7.x issue, but would break any other workloads > SpringBoot 2.7.x and still utilizing the previous version of the mysql r2dbc driver.

Reply:
>1. The processor has already set precedent of checking for the presence of something on the classpath...

I'm not a fan of that either. Spring Boot can deduce the driver from the URI, so I feel that by doing this we are making the code less flexible and we're overriding the Boot default behavior unnecessarily, which could be unexpected for users. 

It is also harder to test since it is tied to what is loaded from the classpath (probably why there's no test that the alternative JDBC driver is set).

>2. Setting the DB protocol mysql vs mariadb at the service binding feels like the wrong direction of control. The service binding should NOT be aware of this level of application implementation detail.

I disagree. I think a service provider would know if it's a MySQL database or a MariaDb database since it's providing that service. It probably has the best understanding of what client drivers should be used to talk to the database. As a library author, we can only try to make guesses and assumptions based on what's visible at the time we run. IMHO, that's less accurate.

It's also already setting the "type" which indicates "mysql". It is tempting to want to use the type as the connection protocol (or if the types don't match exactly as a key to map to the protocol, i.e. if type == x, use 'mysql'. if type == y, use 'mariadb').

> In what field of the service binding would you put the protocol? MySQL Operators like the VMWare Tanzu SQL only include attributes such as port, host, username/password. Including a the DB protocol would require either the Operator to be updated or the field to added manually by another operator persona.

I was thinking field initially, but the type could work. It is contextually similar and 1/2 of the types being used are already correct, i.e. `mysql` is being used now and that maps to the protocol already. It would just be a matter of service providers that are

Reply:
>What initially drove this PR is the fact that Tanzu SQL for MySQL no longer works out of the box (https://github.com/spring-cloud/spring-cloud-bindings/issues/72) (i.e. without addition configuration per MySQL instance) with Spring Boot 2.7.x, and with the solution you suggest, that fact still remains. It may also be true for other MySQL operator offerings that support service bindings (assuming they exist).

I hear what you are saying here. I don't know that this library is the best place to fix this problem though. We could add some magic to try and guess, but it would be best if we're not guessing. The core of this library is just to map values from the binding to well-known Spring Boot properties. If we need a new setting, we should stick to that and take something from the binding and map it to the Boot property. I think anything beyond that is/should be out of scope. I know there are some spots where we're doing that, but I'd be more in favor of pulling that out (can't cause it would be a breaking change) as opposed to propagating that pattern.

Reply:
> I disagree. I think a service provider would know if it's a MySQL database or a MariaDb database since it's providing that service. It probably has the best understanding of what client drivers should be used to talk to the database. As a library author, we can only try to make guesses and assumptions based on what's visible at the time we run. IMHO, that's less accurate.

I think I've failed to articulate the crux of this problem.  

You are correct that the service provider is going to know what type of database is being provided: MySQL in this case based on the `type`.   Being that the service provider has establish MySQL as the database deployed,  I would also partly agree that the service provider "knows" what the best driver should be to connect, however the service provider pragmatically has no control over what is actually available to the application.  In the use case of MySQL and SpringBoot 2.7.x, MySQL would probably be the best driver to use, however this not an option with r2dbc per the SpringBoot documentation which suggests switching to the MariaDB driver to connect to MySQL.  

We're kind of in a weird situation with the latest SpringBoot release as what would probably be deemed the most appropriate database driver is not available and alternative solution doesn't support the URL connection string being served up by spring cloud bindings library.

Frankly I'm not a huge fan of the solution that I have provided, but I have not been able to come up with a better solution at this time.  I have shied away from some the approaches that you have suggested mainly because it feels the application problem at hand is leaking into the operators service binding implementation.

The fact still remains that service binding along with MySQL r2dbc and this library no longer works in the latest release of SpringBoot.  The question is who, where, and how should it be addressed.

Reply:
@dmikusa-pivotal Post our discussion a few weeks ago, has their been any internal discussion on the direction this issue needs to go?

thanks

Reply:
Thanks @gm2552 for your patience. We just discussed and are OK moving forward with this PR.

A few things I'd like to see:

1. I'd like to have the same behavior for the protocol for both jdbc and r2dbc, so [here](https://github.com/spring-cloud/spring-cloud-bindings/blob/main/src/main/java/org/springframework/cloud/bindings/boot/MySqlBindingsPropertiesProcessor.java#L51) should function like what you did [here](https://github.com/spring-cloud/spring-cloud-bindings/pull/73/files#diff-adecc7b789aa0e8fd8d19b9b22e51de0118a799176cbd6c228539a363cdb927bR81).
2. Change [this](https://github.com/spring-cloud/spring-cloud-bindings/#mysql-rdbms) to be mysql or mariadb. In both cases, we support both now.
3. Make a note in the README for the jdbc & r2dbc URL that the protocol is set based on the driver that's on the classpath.
4. Please add a small note at the bottom of the MySQL/MariaDB section that says we examine libraries on the classpath and that having both MySQL & MariaDB JDBC or R2DBC drivers on the classpath is not supported.

Thanks!